More config.toml.example cleanups

jyn514 opened the pull request:
- Link to more documentation
- Move `changelog-seen` into the "Global Settings" section
- Update incorrect comments on `llvm.link-shared` and `rust.debug-assertions`
- Use the correct default in the commented-out example more often
- Clarify that `docs` and `compiler-docs` only control the default, they're not a hard-off switch.
- Document `-vvv` and `local-rebuild`
- Minor improvements to doc-comments in config.toml.example

This also sets `download-rustc = false`; that was already the default, but it will be helpful in case the default changes (https://jyn.dev/2023/01/12/Bootstrapping-Rust-in-2023.html).

albertlarsan68 reviewed (changes requested):
Overall great, just a few nits and questions.

jyn514 commented on config.toml.example:
```diff
@@ -25,6 +19,12 @@ changelog-seen = 2
 
+# Keeps track of the last version of `x.py` used.
+# If it does not match the version that is currently running,
+# `x.py` will prompt you to update it and read the changelog.
```
It reads `changelog-seen`, and prompts you to read `CHANGELOG.md`.

albertlarsan68 replied:
```suggestion
# `x.py` will prompt you to update it and to read the changelog.
```
Not a native English speaker, but it seems more clear this way.

jyn514 commented on config.toml.example:
```diff
@@ -361,10 +373,10 @@ changelog-seen = 2
 
-# Instead of installing to /usr/local, install to this path instead.
+# Where to install the generated toolchain.
 #prefix = "/usr/local"
```
No. The paths below are relative to this directory. I've updated the comment.

jyn514 commented on config.toml.example:
```diff
@@ -418,9 +430,9 @@ changelog-seen = 2
+# This is mostly useful for tools; if you have changes to `compiler/` or `library/` they will be ignored.
 #
-# You can set this to "if-unchanged" to only download if `compiler/` has not been modified.
+# You can set this to "if-unchanged" to only download if the compiler and standard library have not been modified.
```
No. `true` downloads the compiler unconditionally, even if there are changes.

albertlarsan68 replied:
Is is also the case with `download-ci-llvm`?

jyn514 replied:
Yes.

jyn514 replied:
Actually `download-ci-llvm` downloads unconditionally even if you set it to `if-available`; we might want to improve that at some point but right now we don't even rebuild on changes, so I don't think it's the most pressing issue.

jyn514 commented on src/bootstrap/defaults/config.user.toml:
```diff
@@ -9,5 +9,6 @@ doc-stage = 2
-# Most users installing from source want to build all parts of the project from source, not just rustc itself.
+# Most users installing from source want to build all parts of the project from source.
 download-ci-llvm = false
+download-rustc = false
```
It definitely does not :sweat_smile: good catch, ty!

jyn514 commented on config.toml.example:
```diff
@@ -584,12 +591,16 @@ changelog-seen = 2
 # Having the git information can cause a lot of rebuilds during development.
 # Note: If this attribute is not explicitly set (e.g. if left commented out) it
```
I've removed this line entirely, I think the line with the default below is clear enough.

jyn514 committed:
More config.toml.example cleanups

- Link to more documentation
- Move `changelog-seen` into the "Global Settings" section
- Update incorrect comments on `llvm.link-shared` and
  `rust.debug-assertions`
- Use the correct default in the commented-out example more often
- Clarify that `docs` and `compiler-docs` only control the default,
  they're not a hard-off switch.
- Document `-vvv` and `local-rebuild`
- Minor improvements to doc-comments in config.toml.example

This also sets `download-rustc = false`; that was already the default,
but it will be helpful in case the default changes
(https://jyn.dev/2023/01/12/Bootstrapping-Rust-in-2023.html).

albertlarsan68 commented on config.toml.example:
```diff
@@ -185,47 +182,56 @@ changelog-seen = 2
 # host triples to this list as well in order for those host toolchains to be
 # able to compile programs for their native target.
-#target = ["x86_64-unknown-linux-gnu"] (as an example)
+#target = [host]
```
```suggestion
#target = host
```

Since `host` is already a list

jyn514 replied:
Sure, done. I also added a comment about what values are accepted here since it's kinda misleading that `host` is singular.

albertlarsan68 commented on config.toml.example:
```diff
@@ -418,9 +430,10 @@ changelog-seen = 2
-# You can set this to "if-unchanged" to only download if `compiler/` has not been modified.
+# Set this to "if-unchanged" to only download if the compiler and standard library have not been modified.
+# Set this to `true` to download unconditionally (useful if e.g. you are only changing doc-comments).
 #download-rustc = false
```
Also, does the test suite passes with this option?

jyn514 replied:
All but two of the UI tests pass with this enabled. I opened https://github.com/rust-lang/rust/issues/108767 to track the two failing tests.

About half the debuginfo tests fail because of https://github.com/rust-lang/simpleinfra/issues/254, I haven't bothered to get them working but I don't think it's related to this option.

The standard library unit tests fail: https://github.com/rust-lang/rust/issues/108768

`tests/run-make/translation` fails, but I'm not sure if it's related to this change or not:
```
error: failed to load fluent bundle: failed to add resource: Attempt to override an existing message: "parse_struct_literal_body_without_path".
```

albertlarsan68 commented on config.toml.example:
```diff
@@ -185,47 +182,56 @@ changelog-seen = 2
 #
-# Defaults to platform where `x.py` is run.
+# If `rustc` is present in path, this defaults to the host it was compiled for.
+# Otherwise, `x.py` will try to infer it from the output of `uname`.
```
What happens on Windows?

jyn514 replied:
We hard-code MSVC: https://github.com/rust-lang/rust/blob/1057e2132e32764676e89e88992dc7a4c3e955ea/src/bootstrap/bootstrap.py#L226-L232

albertlarsan68 replied:
Does msys2/Whatever-is-the-main-environnement-used-to-build-the-`*-pc-windows-gnu`-toolchains have `uname` working?

jyn514 replied:
I don't know. I would rather not block this PR on fixing all of bootstrap's bugs.

albertlarsan68 replied:
I read the code a bit below, and it seems like it can detect the presence of `uname` and use it, but I think adding a note that when `uname` does not exist/does not return an expected value on Windows, it defaults to MSVC.

albertlarsan68 commented on config.toml.example:
```diff
@@ -185,47 +182,56 @@ changelog-seen = 2
 # friends will still be compiled but they will not be used to generate any
 # documentation.
+#
+# You can still build documentation when this is disabled by explicitly passing paths, e.g. `x doc
+# library`.
```
```suggestion
# You can still build documentation when this is disabled by explicitly passing paths, e.g. 
# `x doc library`.
```

albertlarsan68 commented:
There is an inconsistency:
Sometimes, to say that a value is the same as another, they are sometimes referred by their full path and sometimes only by their names.

jyn514 commented:
Ok, fixed.

Can you please switch the labels to S-waiting-on-author after you give feedback? It makes it easier for me to find the PR.

bors commented:
:umbrella: The latest upstream changes (presumably #109056) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jyn514 commented:
I am not planning to rebase this until I get a review - let me know if that makes it difficult, but the conflict is in config.user.toml so it shouldn't be too bad.

albertlarsan68 reviewed (approved):
Thanks for the PR!
@bors rollup
r=me once rebased

jyn514 commented:
@bors r=albertlarsan68
